thunder/tests/test_grad.py

```diff
@@ -258,15 +258,29 @@ def _replace_none_with_zero(x, y):
     x = list(x)
     y = list(y)
     assert x[0] is not None or y[0] is not None, "Both x and y are None"
-    device = x[0].device if x[0] is not None else y[0].device
-    zero = torch.tensor(0.0, device=device, dtype=torch.float64)
     for i, (a, b) in enumerate(zip(x, y)):
         if a is None or b is None:
-            x[i] = zero
-            y[i] = zero
+            device = x[i].device if x[i] is not None else y[i].device
+            x[i] = torch.tensor(0.0, device=device, dtype=torch.float64)
+            y[i] = torch.tensor(0.0, device=device, dtype=torch.float64)
+
     return x, y
 
 
+def _is_cpu_scalar_tensor(x):
+    return isinstance(x, torch.Tensor) and x.device.type == "cpu" and x.ndim == 0
+
+
+# If one tensor is a CPU scalar tensor and the other is on CUDA, move the scalar tensor to CUDA
+# Then do the ravel and dot operation
+def _tensor_dot(x, y):
+    if _is_cpu_scalar_tensor(x) and y.is_cuda:
```

Comment:
Can this call the function defined in utils instead of defining a new one?

Reply:
I modified the function in utils to detect cpu scalar tensor for both types torch.Tensor and TensorProxy (I think it's no harm to also check torch.Tensor in utils in this case?), so then we can also use it here